In [1]:
import os
import glob
import numpy as np
import pandas as pd

def naive_bayes_train(X, y):
    # Calcular las probabilidades previas de cada clase
    labels, label_counts = np.unique(y, return_counts=True)
    prior_probs = label_counts / len(y)
    priors = dict(zip(labels, prior_probs))  # Mapea cada etiqueta con su probabilidad previa

    # Calcular las probabilidades condicionales para cada atributo dado cada clase
    cond_probs = {}
    for label in labels:
        subset = X[y == label]  # Se filtra los datos para la clase actual
        features_probs = {}
        for col in X.columns:
            levels, counts = np.unique(subset[col], return_counts=True)
            probs = counts / counts.sum()  # Probabilidad de cada valor del atributo dentro de la clase
            features_probs[col] = dict(zip(levels, probs))
        cond_probs[label] = features_probs  # Se mapea la clase con las probabilidades de sus atributos

    return priors, cond_probs

def naive_bayes_predict(X, priors, cond_probs):
    predictions = []
    for _, row in X.iterrows():  # Iterar sobre cada instancia para hacer predicciones
        label_probs = {}
        for label, features_probs in cond_probs.items():
            total_prob = np.log(priors[label])
            for feature, value in row.items():
                feature_probs = features_probs.get(feature, {})
                prob = feature_probs.get(value, 1e-6)  # Laplace smoothing para valores no vistos
                total_prob += np.log(prob)
            label_probs[label] = total_prob
        predictions.append(max(label_probs, key=label_probs.get))  # Escoger la clase con mayor probabilidad
    return predictions

def simple_cross_validate(data, target, k=10):
    # Mezclamos y separamos en k particiones
    folds = np.array_split(data.sample(frac=1, random_state=42), k)
    accuracies = []

    for i in range(k):
        train = pd.concat([folds[j] for j in range(k) if j != i])
        test = folds[i]
        priors, cond_probs = naive_bayes_train(train.drop(target, axis=1), train[target])
        predictions = naive_bayes_predict(test.drop(target, axis=1), priors, cond_probs)
        accuracy = np.mean(predictions == test[target])
        accuracies.append(accuracy)

    return np.mean(accuracies), np.std(accuracies)

# Ruta de la carpeta que contiene las bases de datos (archivos CSV)
carpeta_bd = r"C:\Users\Carlo\Desktop\IA\MDLP\base de datos discretizadas con mdlp R"

# Buscar todos los archivos CSV en la carpeta
archivos_csv = glob.glob(os.path.join(carpeta_bd, "*.csv"))

# Lista para almacenar los resultados
resultados = []

for archivo in archivos_csv:
    try:
        # Cargar la base de datos
        data = pd.read_csv(archivo)
        
        # Verificamos que la columna objetivo 'class' exista
        if 'class' not in data.columns:
            print(f"El archivo {os.path.basename(archivo)} no contiene la columna 'class'. Se omite.")
            continue
        
        # Ejecutar la validación cruzada simple
        mean_accuracy, std_deviation = simple_cross_validate(data, 'class')
        
        # Agregar los resultados a la lista
        resultados.append({
            "Dataset": os.path.basename(archivo),
            "Mean Accuracy": mean_accuracy,
            "Std Deviation": std_deviation
        })
        print(f"Procesado {os.path.basename(archivo)} - Accuracy: {mean_accuracy:.4f} / Std: {std_deviation:.4f}")
    except Exception as e:
        print(f"Error al procesar {os.path.basename(archivo)}: {e}")

# Crear un DataFrame con los resultados
df_resultados = pd.DataFrame(resultados)
print("\nResultados finales:")
print(df_resultados)

c:\Users\Carlo\AppData\Local\Programs\Python\Python312\Lib\site-packages\numpy\core\fromnumeric.py:59: FutureWarning: 'DataFrame.swapaxes' is deprecated and will be removed in a future version. Please use 'DataFrame.transpose' instead.
  return bound(*args, **kwds)


Procesado dry_bean_mdlpR.csv - Accuracy: 0.8998 / Std: 0.0051
Procesado glass_mdlpR.csv - Accuracy: 0.9905 / Std: 0.0190
Procesado iris_mdlpR.csv - Accuracy: 0.9333 / Std: 0.0667


c:\Users\Carlo\AppData\Local\Programs\Python\Python312\Lib\site-packages\numpy\core\fromnumeric.py:59: FutureWarning: 'DataFrame.swapaxes' is deprecated and will be removed in a future version. Please use 'DataFrame.transpose' instead.
  return bound(*args, **kwds)
c:\Users\Carlo\AppData\Local\Programs\Python\Python312\Lib\site-packages\numpy\core\fromnumeric.py:59: FutureWarning: 'DataFrame.swapaxes' is deprecated and will be removed in a future version. Please use 'DataFrame.transpose' instead.
  return bound(*args, **kwds)
c:\Users\Carlo\AppData\Local\Programs\Python\Python312\Lib\site-packages\numpy\core\fromnumeric.py:59: FutureWarning: 'DataFrame.swapaxes' is deprecated and will be removed in a future version. Please use 'DataFrame.transpose' instead.
  return bound(*args, **kwds)


Procesado letter_recognition_mdlpR.csv - Accuracy: 0.7480 / Std: 0.0074
Procesado rice_mdlpR.csv - Accuracy: 0.9184 / Std: 0.0082


c:\Users\Carlo\AppData\Local\Programs\Python\Python312\Lib\site-packages\numpy\core\fromnumeric.py:59: FutureWarning: 'DataFrame.swapaxes' is deprecated and will be removed in a future version. Please use 'DataFrame.transpose' instead.
  return bound(*args, **kwds)
c:\Users\Carlo\AppData\Local\Programs\Python\Python312\Lib\site-packages\numpy\core\fromnumeric.py:59: FutureWarning: 'DataFrame.swapaxes' is deprecated and will be removed in a future version. Please use 'DataFrame.transpose' instead.
  return bound(*args, **kwds)


Procesado seeds_mdlpR.csv - Accuracy: 0.9190 / Std: 0.0372


c:\Users\Carlo\AppData\Local\Programs\Python\Python312\Lib\site-packages\numpy\core\fromnumeric.py:59: FutureWarning: 'DataFrame.swapaxes' is deprecated and will be removed in a future version. Please use 'DataFrame.transpose' instead.
  return bound(*args, **kwds)


Procesado winequality-red_mdlpR.csv - Accuracy: 0.5785 / Std: 0.0378


c:\Users\Carlo\AppData\Local\Programs\Python\Python312\Lib\site-packages\numpy\core\fromnumeric.py:59: FutureWarning: 'DataFrame.swapaxes' is deprecated and will be removed in a future version. Please use 'DataFrame.transpose' instead.
  return bound(*args, **kwds)


Procesado winequality-white_mdlpR.csv - Accuracy: 0.4886 / Std: 0.0152


c:\Users\Carlo\AppData\Local\Programs\Python\Python312\Lib\site-packages\numpy\core\fromnumeric.py:59: FutureWarning: 'DataFrame.swapaxes' is deprecated and will be removed in a future version. Please use 'DataFrame.transpose' instead.
  return bound(*args, **kwds)


Procesado yeast_mdlpR.csv - Accuracy: 0.5917 / Std: 0.0390

Resultados finales:
                        Dataset  Mean Accuracy  Std Deviation
0            dry_bean_mdlpR.csv       0.899786       0.005069
1               glass_mdlpR.csv       0.990476       0.019048
2                iris_mdlpR.csv       0.933333       0.066667
3  letter_recognition_mdlpR.csv       0.748050       0.007353
4                rice_mdlpR.csv       0.918373       0.008170
5               seeds_mdlpR.csv       0.919048       0.037192
6     winequality-red_mdlpR.csv       0.578487       0.037823
7   winequality-white_mdlpR.csv       0.488571       0.015185
8               yeast_mdlpR.csv       0.591679       0.039013


Resultados finales de mi mdlp:
                       Dataset  Mean Accuracy  Std Deviation
0            dry_bean_mdlp.csv       0.900521       0.004895
1               gamma_mdlp.csv       0.782019       0.007418
2               glass_mdlp.csv       0.967100       0.030300
3                iris_mdlp.csv       0.673333       0.113333
4  letter_recognition_mdlp.csv       0.745450       0.006725
5          rice_cameo_mdlp.csv       0.918373       0.008170
6               seeds_mdlp.csv       0.919048       0.042857
7     winequality-red_mdlp.csv       0.583502       0.045573
8   winequality-white_mdlp.csv       0.488366       0.014111
9               yeast_mdlp.csv       0.588296       0.028863

Resultados finales con mi CAIM:

Resultados finales:
               Dataset  Mean Accuracy  Std Deviation
0    dry-bean_caim.csv       0.892293       0.006950
1       gamma_caim.csv       0.738275       0.008083
2       glass_caim.csv       0.701948       0.093185
3        iris_caim.csv       0.933333       0.051640
4      letter_caim.csv       0.748100       0.008752
5        rice_caim.csv       0.914436       0.009694
6       seeds_caim.csv       0.880952       0.048795
7    wine-red_caim.csv       0.568502       0.036518
8  wine-white_caim.csv       0.513683       0.018552
9       yeast_caim.csv       0.568751       0.042036

Resultados finales con mdlp R:

Resultados finales:
                        Dataset  Mean Accuracy  Std Deviation
0            dry_bean_mdlpR.csv       0.899786       0.005069
1               glass_mdlpR.csv       0.990476       0.019048
2                iris_mdlpR.csv       0.933333       0.066667
3  letter_recognition_mdlpR.csv       0.748050       0.007353
4                rice_mdlpR.csv       0.918373       0.008170
5               seeds_mdlpR.csv       0.919048       0.037192
6     winequality-red_mdlpR.csv       0.578487       0.037823
7   winequality-white_mdlpR.csv       0.488571       0.015185
8               yeast_mdlpR.csv       0.591679       0.039013